compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -595,6 +571,115 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
             }
         }
     }
+
+    /// Creates an incomplete operand containing the [`abi::Scalar`]s expected based
+    /// on the `layout` passed. This is for use with [`OperandRef::insert_field`]
+    /// later to set the necessary immediate(s).
+    ///
+    /// ICEs on types for which [`OperandRef::supports_builder`] returns false.
+    pub(crate) fn builder(layout: TyAndLayout<'tcx>) -> OperandRef<'tcx, Result<V, abi::Scalar>> {
+        let val = match layout.backend_repr {
+            BackendRepr::Memory { .. } if layout.is_zst() => OperandValue::ZeroSized,
+            BackendRepr::Scalar(s) => OperandValue::Immediate(Err(s)),
+            BackendRepr::ScalarPair(a, b) => OperandValue::Pair(Err(a), Err(b)),
+            _ => bug!("Cannot use type in operand builder: {layout:?}"),
+        };
+        OperandRef { val, layout }
+    }
+
+    /// Determines whether [`OperandRef::builder`] is supported for this `layout`.
+    ///
+    /// Only needed by [`FunctionCx::rvalue_creates_operand`], but here so the
+    /// two `match`es are proximate in implementation.
+    pub(crate) fn supports_builder(layout: TyAndLayout<'tcx>) -> bool {
+        match layout.backend_repr {
+            BackendRepr::Memory { .. } if layout.is_zst() => true,
+            BackendRepr::Scalar(_) | BackendRepr::ScalarPair(_, _) => true,
+            BackendRepr::Memory { .. } | BackendRepr::SimdVector { .. } => false,
+        }
+    }
+}
```

Comment:
Right, I believe my main question here is why couldn't we just have a return of `Option<OperandRef<...>>` to prevent them from getting out of sync?

Reply:
Obviously that requires skipping the `bug!`

Reply:
Yeah, and even if `supports_builder` is a separate function it will probably generate the code just fine if it relies on LLVM optimizing `builder(layout).is_ok()` or something.